src/dos/cdrom.cpp

```diff
@@ -18,6 +18,217 @@
+{
+	StopAudio();
+	SDL_CDClose(cd);
+	cd = 0;
+}
+
+bool CDROM_Interface_SDL::SetDevice(char *path, int forceCD)
+{
+	char buffer[512];
+	strcpy(buffer, path);
+	upcase(buffer);
+
+	int num = SDL_CDNumDrives();
+	if ((forceCD >= 0) && (forceCD < num)) {
+		driveID = forceCD;
+		cd = SDL_CDOpen(driveID);
+		SDL_CDStatus(cd);
+		return true;
+	};
+
+	const char *cdname = 0;
+	for (int i = 0; i < num; i++) {
+		cdname = SDL_CDName(i);
+		if (strcmp(buffer, cdname) == 0) {
+			cd = SDL_CDOpen(i);
+			SDL_CDStatus(cd);
+			driveID = i;
+			return true;
+		};
+	};
+	return false;
+}
+
+bool CDROM_Interface_SDL::GetAudioTracks(uint8_t &stTrack, uint8_t &end, TMSF &leadOut)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		stTrack = 1;
+		end = cd->numtracks;
+		leadOut = frames_to_msf(cd->track[cd->numtracks].offset);
+	}
+	return CD_INDRIVE(SDL_CDStatus(cd));
+}
+
+bool CDROM_Interface_SDL::GetAudioTrackInfo(uint8_t track,
+                                            TMSF &start,
+                                            unsigned char &attr)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		start = frames_to_msf(cd->track[track - 1].offset);
+		attr = cd->track[track - 1].type << 4; // sdl uses 0 for audio
+		                                       // and 4 for data. instead
+		                                       // of 0x00 and 0x40
+	}
+	return CD_INDRIVE(SDL_CDStatus(cd));
+}
+
+bool CDROM_Interface_SDL::GetAudioSub(unsigned char &attr,
+                                      unsigned char &track,
+                                      unsigned char &index,
+                                      TMSF &relPos,
+                                      TMSF &absPos)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		track = cd->cur_track;
+		index = cd->cur_track;
+		attr = cd->track[track].type << 4;
+		relPos = frames_to_msf(cd->cur_frame);
+		absPos = frames_to_msf(cd->cur_frame + cd->track[track].offset);
+	}
+	return CD_INDRIVE(SDL_CDStatus(cd));
+}
+
+bool CDROM_Interface_SDL::GetAudioStatus(bool &playing, bool &pause)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		playing = (cd->status == CD_PLAYING);
+		pause = (cd->status == CD_PAUSED);
+	}
+	return CD_INDRIVE(SDL_CDStatus(cd));
+}
+
+bool CDROM_Interface_SDL::GetMediaTrayStatus(bool &mediaPresent,
+                                             bool &mediaChanged,
+                                             bool &trayOpen)
+{
+	SDL_CDStatus(cd);
+	mediaPresent = (cd->status != CD_TRAYEMPTY) && (cd->status != CD_ERROR);
+	mediaChanged = (oldLeadOut != cd->track[cd->numtracks].offset);
+	trayOpen = !mediaPresent;
+	oldLeadOut = cd->track[cd->numtracks].offset;
+	if (mediaChanged)
+		SDL_CDStatus(cd);
+	return true;
+}
+
+bool CDROM_Interface_SDL::PlayAudioSector(const uint32_t start, uint32_t len)
+{
+	// Has to be there, otherwise wrong cd status report (dunno why, sdl bug ?)
+	SDL_CDClose(cd);
+	cd = SDL_CDOpen(driveID);
+	bool success = (SDL_CDPlay(cd, start + 150, len) == 0);
+	return success;
```

Comment:
Direct return.

Reply:
The success looks redundant, but it makes the evaluation self-documenting.

The problem is this is mixing the old C-style "zero-is good" with booleans, where zero-is-bad.

So a direct return would work if we were C-through-and-through:

`return SDL_CDPlay(cd, start + 150, len)`

Then the we'd value the return `int` as 0 and declare success.

So a direct return becomes `SDL_CDPlay(cd, start + 150, len) == 0`

And without any indication of success of fail, it's harder to trust - because the API itself is using `bool` and not int.

Reply:
I was actually meaning `return (SDL_CDPlay(cd, start + 150, len) == 0);`, not `return SDL_CDPlay(cd, start + 150, len);`

Basically, keeping the test in.

Reply:
I suppose that's clear enough; the test describes the criteria.


``` c++
Return !SDL_CDPlay(cd, start + 150, len);
```
(Just kidding!)

Reply:
I changed it to “return (SDL_CDPlay(cd, start + 150, len) == 0);” accordingly.